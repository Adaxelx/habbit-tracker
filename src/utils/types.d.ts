export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
