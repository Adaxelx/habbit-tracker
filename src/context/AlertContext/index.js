export { default, useAlertContext } from './AlertContext';
