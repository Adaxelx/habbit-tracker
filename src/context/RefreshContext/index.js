export { default, useRefreshContext } from './RefreshContext';
