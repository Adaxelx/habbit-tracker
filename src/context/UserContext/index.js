export { default, useUserContext, UserContext } from './UserContext';
