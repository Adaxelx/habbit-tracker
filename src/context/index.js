export { default as ThemeProvider, useThemeContext } from './ThemeContext';
export { default as UserProvider, useUserContext, UserContext } from './UserContext';
export { default as AlertProvider, useAlertContext, AlertContext } from './AlertContext';
export { default as RefreshProvider, useRefreshContext } from './RefreshContext';
