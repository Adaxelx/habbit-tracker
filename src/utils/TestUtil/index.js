export { default } from './TestUtil.ts';
