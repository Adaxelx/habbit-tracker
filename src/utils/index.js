export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
