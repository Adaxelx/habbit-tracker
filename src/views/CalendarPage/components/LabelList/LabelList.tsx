import React from 'react';
import { useUserContext, useRefreshContext } from 'context';
import { PopUp } from 'components';
import { FormWithLabels } from 'utils';
import { deleteLabel } from 'views/CalendarPage/CalendarPage.api';
import { useMutation } from 'hooks';
import { Label } from '..';

const LabelList = ({ handleClose, open, labels }: FormWithLabels) => {
  const { token } = useUserContext();

  const { handleRefHabbit, handleRefLabel } = useRefreshContext();

  const [mutate, loading] = useMutation({
    request: (id: string) => deleteLabel(token, id),
    refresh: [handleRefLabel, handleRefHabbit],
    messageSuccess: 'Succesfuly deleted label.',
  });

  const handleDelete = async (id: string) => mutate(id);

  return (
    <PopUp open={open} handleClose={handleClose} header="Label list" disabled={loading}>
      <>
        {labels?.map((label) => (
          <Label label={label} key={label._id} disabled={loading} handleDelete={handleDelete} />
        ))}
      </>
    </PopUp>
  );
};

export default LabelList;
