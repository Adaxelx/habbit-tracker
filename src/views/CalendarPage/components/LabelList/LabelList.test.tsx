import React from 'react';
import 'jest-styled-components';
import { waitFor } from '@testing-library/react';
import { AlertTypes, FormWithLabels, TestUtil } from 'utils';
import { AlertContext } from 'context';
import LabelList from './LabelList';

describe('LabelList', () => {
  let util: TestUtil;
  let props: FormWithLabels;
  let fail: boolean;
  let alertC = { showAlert: jest.fn() };

  const mockedFetch = () => {
    if (fail) {
      return Promise.resolve({
        status: 500,
        ok: false,
        json: () => Promise.resolve({ message: 'error' }),
      });
    }

    return Promise.resolve({
      status: 204,
      ok: true,
    });
  };

  beforeAll(() => {
    jest.spyOn(window, 'fetch');
  });

  beforeEach(() => {
    props = {
      labels: [{ _id: '1', title: 'title', color: '#333333', userId: '123' }],
      open: true,
      handleClose: jest.fn(),
    };
    jest.clearAllMocks();
    fail = false;
    util = new TestUtil(
      (
        <AlertContext.Provider value={alertC}>
          <LabelList {...props} />
        </AlertContext.Provider>
      ),
    );
    window.fetch.mockImplementation(mockedFetch);
  });

  it('should match snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('should delete label', async () => {
    util.click('delete1');

    await waitFor(() => expect(util.get('deleteModal-close')).toBeInTheDocument());
    util.click('delete-keep');
    util.click('delete1');
    await waitFor(() => expect(util.get('deleteModal-close')).toBeInTheDocument());
    util.click('deleteItem');
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('Succesfuly deleted label.', AlertTypes.SUCCESS);
  });

  it('should not delete label if api fails', async () => {
    fail = true;
    util.click('delete1');

    await waitFor(() => expect(util.get('deleteModal-close')).toBeInTheDocument());
    util.get('deleteModal-close');
    util.click('delete1');
    await waitFor(() => expect(util.get('deleteModal-close')).toBeInTheDocument());
    util.click('deleteItem');
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    expect(alertC.showAlert).toHaveBeenCalledWith('error');
  });
});
