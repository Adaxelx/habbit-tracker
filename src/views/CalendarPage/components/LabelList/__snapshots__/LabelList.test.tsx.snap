// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelList should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1000;
  background-color: rgba(0,0,0,0.3);
}

<div
    class="c0"
  />
  .c9 {
  background-color: #333333;
  padding: 0.75rem;
  width: auto;
  border-radius: 20px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 0.75rem;
  padding-top: 0;
}

.c6 {
  margin-bottom: 0.5rem;
}

.c7 {
  color: #fff;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: none;
  display: inline-block;
  text-align: center;
  margin-top: 0;
  margin-left: 0;
  margin-right: 0.75rem;
}

.c7:hover {
  cursor: pointer;
}

.c7:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c8 {
  color: #000000;
  background-color: #FF9191;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: none;
  display: inline-block;
  text-align: center;
  margin-top: 0;
  margin-left: 0;
  margin-right: 0;
}

.c8:hover {
  cursor: pointer;
}

.c8:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c0 {
  position: fixed;
  width: calc(100% - 1rem - 1rem);
  height: calc(90vh - 1rem - 1rem);
  top: 1rem;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  padding: 1rem 0;
  background-color: #AFE9FF;
  border: 1px solid #707070;
  z-index: 1000;
  max-width: 350px;
  overflow: scroll;
}

.c4 {
  width: 100%;
  text-align: center;
  font-size: 2rem;
}

.c1 {
  width: 50px;
  height: 50px;
  position: relative;
  display: block;
  margin-left: auto;
  margin-right: 0;
  border: none;
  background-color: transparent;
}

.c1:hover {
  cursor: pointer;
}

.c1:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c2 {
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0%;
  left: 50%;
  background-color: #FF9191;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c3 {
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0%;
  left: 50%;
  background-color: #FF9191;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="Label list-close"
      type="button"
    >
      <div
        class="c2"
        rotate="45"
      />
      <div
        class="c3"
        rotate="-45"
      />
    </button>
    <h2
      class="c4"
      data-testid="Label list-header"
    >
      Label list
    </h2>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <button
          class="c7"
          data-testid="edit"
        >
          Edit label
        </button>
        <button
          class="c8"
          data-testid="delete1"
        >
          X
        </button>
      </div>
      <small
        class="c9"
        color="#333333"
      >
        title
      </small>
    </div>
  </div>
</DocumentFragment>
`;
